Addition of BpyBuild API

StandingPadAnimations opened the pull request:
This PR tracks the creation of a basic API for developers to use in BpyBuild

StandingPadAnimations commented:
@TheDuckCow this is ready for review now

TheDuckCow commented on test/default.py:
```diff
@@ -1 +1,25 @@
+    print(os.listdir())
+
+
+def post_install() -> None:
```
Could we have one or two args? At least we need the path of the folder we just installed to, so I'm able to do the renaming of folders in the target location.

StandingPadAnimations replied:
I've added a `BabContext` object that provides the “intended working directory” (i.e., what the working directory should be):
```py
from bpy_addon_build.api import BabContext

def main(ctx: BabContext) -> None:
	print(ctx.current_path)
```

TheDuckCow commented on bpy_addon_build/build_context/hook_definitions.py:
```diff
@@ -0,0 +1,160 @@
+        if action not in ctx.api.action_mods:
+            print("Action not in API!")
+            return
+        if hasattr(ctx.api.action_mods[action], "prebuild"):
```
```suggestion
        if hasattr(ctx.api.action_mods[action], "pre_build"):
```

Per your example file below, I think this needs to be `pre_build` instead of "prebuild", as it doesn't print "HELLO"

TheDuckCow commented on bpy_addon_build/build_context/hook_definitions.py:
```diff
@@ -0,0 +1,160 @@
+    Returns:
+        None
+    """
+    if ctx.config.build_actions is None and len(ctx.cli.actions):
```
Minor style comment that here and below, the nesting is getting pretty deep. Would apply to the other hooks.

I'm a fan of early returns, so you could have basically only ever one level of indentnation.

e.g. it starts with
```
if not if action in ctx.cli.actions:
    return
if action not in ctx.api.action_mods:
    print("Action not in API!")
    return
if not hasattr(ctx.api.action_mods[action], "post_install"):
    return
...

TheDuckCow commented on docs/actions.md:
```diff
@@ -0,0 +1,63 @@
+- `main`: executed during the build process; working directory is set to a copy of the source tree under `build/stage-1`
+- `pre_build`: executed before building; working directory is set to the `addon_folder` variable defined in `bpy-build.yaml`
+- `pre_install`: executed before installing the built addon; working directory is set to `build/`
+- `post_install`: executed after installing the built addon; working directory is set to the addons folder of the Blender version last installed to
```
I mention below the request to get in the path of the install location, but I suppose it could also work if as described above. However, I'm not finding this to be the case in practice, with this code:

```python
def post_install() -> None:
    print("POST INSTALL")
    print(os.getcwd())
```

I'm getting this print out (ie my github folder loc)

```
POST INSTALL
/Users/patrickcrawford/Documents/gits/mcprep
```

StandingPadAnimations replied:
Works fine on my end:
```
Installed to /home/mahid/.config/blender/3.5/scripts/addons
POST INSTALL
/home/mahid/.config/blender/3.5/scripts/addons
Installed to /home/mahid/.config/blender/3.4/scripts/addons
POST INSTALL
/home/mahid/.config/blender/3.4/scripts/addons
Installed to /home/mahid/.config/blender/2.80/scripts/addons
POST INSTALL
/home/mahid/.config/blender/2.80/scripts/addons
```

TheDuckCow reviewed (changes requested):
Nice work, and nice design/separation of components.

Change request is for either ensuring the cwd is actually changed to match the docs, or to update the args of the function hooks to have the path in question personally, I think that might actually be nicer anyways, but I'll leave that choice up to you. Downside of that would be deciding the name of the path arg input to each hook which would vary by context, but then documentation reinforces what it is which is nice).

StandingPadAnimations commented:
Since BpyBuild expects certain types for hooks, I've gone ahead and added `typeguard` as a dependency so we can check types. In the future, however, I plan to move away from this dependency so we can get easier to digest errors for developers.

TheDuckCow commented on bpy_addon_build/build_context/hook_definitions.py:
```diff
@@ -19,24 +24,44 @@ def build_action_prebuild(ctx: BuildContext, action: str, console: Console) -> N
+    if action not in ctx.cli.actions:
+        return
+    if action not in ctx.api.action_mods:
+        print("Action not in API!")
```
I'm also a bit confused while trying to test this. I'm basically _always_ getting the action is not found, because the "action" is just matching the passed in variable name, as opposed to the hook name.

```
if action not in ctx.api.action_mods:
        print("Action pre_install not in API!", action)
        return
```

The actual printout line I get is `Action pre_install not in API! fast`

The line I'm compiling with is `bab -b fast -v 4.0` where `fast` corresponds to a `fast.py` file. What I was expecting is that it would call the `pre_install` hook

Maybe I'll understand better if we already did the implementation of the fast command, ie the ignore pattern for MCprep_resources paired with pre_install moving the installed MCprep_resources folder to a temp place, and then putting it back in the post_install step. That's all I'm hoping to accomplish here!

My full fast.py snippet:

```python
from bpy_addon_build.api import BabContext

print("Did this run...? fast")  # this seems to never print


def main(ctx: BabContext) -> None:
    print("Dev build, adding mcprep_dev.txt...")
    with open("mcprep_dev.txt", "w") as f:
        f.write("This file is used to enable debugging mode in MCprep, you can ignore this")


def pre_install(ctx: BabContext) -> None:
    print("HELLO")
    print(ctx)


def post_install(ctx: BabContext) -> None:
    print("POST INSTALL")
    print(ctx)
```

StandingPadAnimations replied:
Pushed some debug information for the API. What does the output look like when passing `-dbg`

TheDuckCow replied:
Welp, TL;DR I forgot we also have to add in this line, so my fast arg wasn't being recognized at all. It all seems working well for me now

```
  fast:
    script: "action-scripts/fast.py"
```

Let me just do a little more of a test before approving this

TheDuckCow reviewed (commented):
At a glance seems like this should be what I want, but still can't quite get it to run my specified function in a fast.py script in my default_scripts folder in MCprep where I have this branch installed. Maybe I'm misunderstanding how this should be working?

TheDuckCow commented:
Something I can already tell I'm going to need in the context variable is the target install location. Seems like the `ctx.current_path` is the build folder within the git folder (at least for me), but I really need the e.g. blender 4.0 addons folder, so I can move the pre-existing MCprep_resources folder away and back after the install.

On a separate note, what's the best way to extend the ignore to skip ever copying or duplicate any files from the MCprep_resources folder? It looks like `ignore_filters.main()` works by having things still copied over once and then deleted after via p.unlink(), but I'm trying to avoid _any_ copying or duplication of the resource folder.

StandingPadAnimations commented:
> Something I can already tell I'm going to need in the context variable is the target install location. Seems like the `ctx.current_path` is the build folder within the git folder (at least for me), but I really need the e.g. blender 4.0 addons folder, so I can move the pre-existing MCprep_resources folder away and back after the install.

This should be given in the `pre_install` hook, so I'll see if I can get that bug resolved

> On a separate note, what's the best way to extend the ignore to skip ever copying or duplicate any files from the MCprep_resources folder? It looks like `ignore_filters.main()` works by having things still copied over once and then deleted after via p.unlink(), but I'm trying to avoid _any_ copying or duplication of the resource folder.

The `pre_build` and `clean_up` hooks, which sets the path to the addon source directory and is executed before BpyBuild copies the source tree and after all installs. You should be able to move files out of the addon folder in `pre_build`, and add them back in `clean_up`

TheDuckCow commented:
Sorry I've been taking ages on this one, I need to test writing my script updates locally to see this is actually doing what I ultimately want it to, would be waste if it didn't. I'll try to get to it soon, just been swamped with too many parallel initiatives.

StandingPadAnimations commented:
> Sorry I've been taking ages on this one, I need to test writing my script updates locally to see this is actually doing what I ultimately want it to, would be waste if it didn't. I'll try to get to it soon, just been swamped with too many parallel initiatives.

No problem, the main issue at the moment is that I can't seem to recreate the issues you're facing locally. I think just to make things cleaner, BpyBuild shouldn't change the working directory and just instead pass the intended directory for the developer to use (which we already do with `BabContext`). It's easier to understand, and will clean up some areas of the codebase (speaking of cleaning up, at some point, I need to figure out how to clean up all the `expanduser()` calls we do in BpyBuild). The issue with that is that it would break backwards compatibility, but that's really a minor thing.

As an aside, I think we could also work on a proper wiki for this, but that's for future me to deal with

StandingPadAnimations commented:
Started work on unittests as well in BpyBuild (although there are some weird areas because this is a CLI program), which should make debugging easier

TheDuckCow commented:
Thanks for that, I keep procrastinating on this one - I'll see if I can get back to live test the baseline functionality this week

StandingPadAnimations commented:
Glad I added build tests, because I forgot to update the test files to work without relying on working directory changes

StandingPadAnimations commented:
@TheDuckCow I've added some extremely basic hook tests that check standard output for the correct paths, except for `post_install` (whose output is system dependent). Tests can be ran with `python -m unittest test.tests` (in a virtual environment with all the dependencies at least, I'll add a `requirements.txt` file as well)

StandingPadAnimations commented on test/tests.py:
```diff
@@ -0,0 +1,175 @@
+        )
+        self.assertIn(f"CLEAN UP {TEST_FOLDER}/test_addon/MCprep_addon", stdout_list)
+
+        # TODO: Check folder of post install
+        #
+        # This is dependent on the system, so
+        # reliably checking this is much harder
+        self.assertRegex(mock_stdout.getvalue(), r"POST INSTALL ")
```
We'll have to figure out a way to test for paths in `post_install`. The `post_install` hook in tests does output the path, but that path differs depending on the system and Blender versions found

StandingPadAnimations replied:
Resolved in [`205226d`](https://github.com/Moo-Ack-Productions/bpy-build/pull/9/commits/205226d56b95a5ac32e014d6aba7a8449860cfa6)

StandingPadAnimations commented on test/tests.py:
```diff
@@ -0,0 +1,175 @@
+        - stage-1 folder
+        - "MAIN" in mock_stdout
+        """
+        with mock.patch(
+            "sys.argv", ["bab", "-c", f"{TEST_FOLDER}/test_addon/bpy-build.yaml"]
+        ):
+            bab.main()
```
As a bit of an off-topic aside, I think I'll add this to the docs as a way of running BpyBuild in tests/scripts

TheDuckCow replied:
I'm all for more documentation - something I realize I'm missing getting back to this is the sample command in the contributing guide on how to run the tests and what setup locally is necessary.

StandingPadAnimations commented:
I've added backwards compatibility to how the `main` function in actins is called. Now, if BpyBuild detects that `main` in an action has no arguments, it will run `main` with the correct working directory set (like how actions are ran in the current release of BpyBuild). This will make it easier to incrementally adopt new updates. As always, I've added a test to make sure it runs properly

TheDuckCow commented:
Oh nice that's a super good addition, thanks for making that. I still need to get to trying this out

StandingPadAnimations commented:
Pushed some tests that involve the MCprep source tree (mostly to make sure BpyBuild remains backwards compatible with MCprep's config). Pretty simple currently, but eventually they'll encompass everything in the MCprep build process.

StandingPadAnimations commented:
And now I've added tests for translation builds with MCprep. One of these tests does perform a patch on an action, but that to test the new hooks feature (since MCprep currently doesn't utilize them)

TheDuckCow commented on test/mcprep-patches/build_pot_prebuild.patch:
Might be good to add into the contrib guide, how to actually make use of this / how this gets used.

StandingPadAnimations replied:
We sort of do. `docs/mcprep-first.md` explains a little bit on these patches; these exist to test new features in MCprep where they would be best used.

TheDuckCow reviewed (approved):
Generally, I'm finding this to work well. I think it's probably better to move ahead, rather than me hold this up further. I had to make a minor diff in the mcprep side code to run tests for MCprep itself (just changing the cmdline call from bpy adodn build to bab).

But, one callout - I finally install poetry for the first time (don't judge too hard), and found it's giving me some errors when following the steps of the main contributing guide; I do still also think it could be nice to include the instructions for running tests, using `python -m unittest discover` me an error since it's not picking up the poetry env (maybe I also just need to learn that better). Still, approving in the meantime.

```
Found 58 errors in 8 files (checked 11 source files)
bpy-build $ poetry run mypy --pretty bpy_addon_build
```

StandingPadAnimations commented:
> Generally, I'm finding this to work well. I think it's probably better to move ahead, rather than me hold this up further. I had to make a minor diff in the mcprep side code to run tests for MCprep itself (just changing the cmdline call from bpy adodn build to bab).

Sounds good. I'll merge this in today and push it to PyPi on the weekend

> But, one callout - I finally install poetry for the first time (don't judge too hard), and found it's giving me some errors when following the steps of the main contributing guide; I do still also think it could be nice to include the instructions for running tests, using `python -m unittest discover` me an error since it's not picking up the poetry env (maybe I also just need to learn that better). Still, approving in the meantime.
> 
> ```
> Found 58 errors in 8 files (checked 11 source files)
> bpy-build $ poetry run mypy --pretty bpy_addon_build
> ```

For tests, you'd run `poetry run python -m unittest discover`, which runs Python in the environment. Alternatively, you could do `poetry shell` to enter the virtual environment like normally

StandingPadAnimations commented:
By the way, if you want to learn more about Poetry, the docs are pretty comprehensive: https://python-poetry.org/docs/
